utils/lazy-pages-fuzzer/src/generate/globals.rs

```diff
@@ -0,0 +1,129 @@
+// This file is part of Gear.
+
+// Copyright (C) 2021-2024 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use anyhow::Result;
+use arbitrary::{Arbitrary, Unstructured};
+
+use gear_wasm_instrument::parity_wasm::{
+    builder,
+    elements::{Instruction, Module},
+};
+
+pub const GLOBAL_NAME_PREFIX: &str = "gear_fuzz_";
+
+#[derive(Debug, Clone)]
+pub struct InjectGlobalsConfig {
+    pub max_global_number: usize,
+    pub max_access_per_func: usize,
+}
+
+impl Default for InjectGlobalsConfig {
+    fn default() -> Self {
+        InjectGlobalsConfig {
+            max_global_number: 10,
+            max_access_per_func: 10,
+        }
+    }
+}
+
+pub struct InjectGlobals<'u> {
+    unstructured: Unstructured<'u>,
+    config: InjectGlobalsConfig,
+}
+
+impl InjectGlobals<'_> {
+    pub fn new(unstructured: Unstructured<'_>, config: InjectGlobalsConfig) -> InjectGlobals<'_> {
+        InjectGlobals {
+            unstructured,
+            config,
+        }
+    }
+
+    pub fn inject<'this>(mut self, mut module: Module) -> Result<(Module, Unstructured<'this>)>
+    where
+        Self: 'this,
+    {
+        let global_names: Vec<_> = ('a'..='z')
+            .take(self.config.max_global_number)
+            .map(|ch| format!("{GLOBAL_NAME_PREFIX}{ch}"))
+            .collect();
+
+        let mut next_global_idx = module.globals_space() as u32;
+
+        let code_section = module
+            .code_section_mut()
+            .ok_or_else(|| anyhow::Error::msg("No code section found"))?;
+
+        // Insert global access instructions
+        for function in code_section.bodies_mut() {
+            let count_per_func = self
+                .unstructured
+                .int_in_range(1..=self.config.max_access_per_func)?;
+
+            for _ in 0..=count_per_func {
+                let array_idx = self.unstructured.choose_index(global_names.len())? as u32;
+                let global_idx = next_global_idx + array_idx;
+
+                let insert_at_pos = self
+                    .unstructured
+                    .choose_index(function.code().elements().len())?;
```

Comment:
Last position in the module must be calculated properly. Look at `SyscallsInvocator::insert_syscalls_into_fn`:

```rust
        // The end of insertion range is second-to-last index, as the last
        // index is defined for `Instruction::End` of the function body.
        // But if there's only one instruction in the function, then `0`
        // index is used as an insertion point.
        let last = if fn_code_len > 1 { fn_code_len - 2 } else { 0 };
```

Reply:
Double checked, so even if it inserts at last position (at position of `End` instruction), `End` instruction will be pushed down for one position and function will be correct.